core/src/main/scala/me/jeffmay/neo4j/client/cypher/Cypher.scala

```diff
@@ -18,50 +25,129 @@ object Cypher {
+    */
+  class DynamicMutableParam private[Cypher](namespace: String)
+    extends Param(namespace) with Dynamic {
+
+    final override def isMutable: Boolean = true
+
+    final override protected def __clsName: String = "ApplyDynamicParam"
+
+    /**
+      * Adds a property value to the resulting [[CypherProps]].
+      *
+      * @note assigning two different values to the same property within a [[CypherStatement]] will result
+      *       in an error.
+      * @return a valid [[CypherParamField]]
+      */
+    def applyDynamic[T](name: String)(value: T)(implicit writer: CypherWrites.Value[T]): CypherArg = {
+      val cypherValue = writer.writes(value)
+      new CypherParamField(namespace, name, cypherValue)
+    }
+  }
+
+  /**
+    * @see [[DynamicImmutableParam]]
+    */
+  def params(namespace: String, props: CypherProps): DynamicImmutableParam = new DynamicImmutableParam(namespace, props)
+
+  /**
+    * Immutable params cannot share the same namespace with other params in the same [[CypherStatement]]
+    * as this would cause confusion if properties were merged and changed from underneath each other.
+    *
+    * @param namespace the namespace to which all parameters inserted by this class share
+    * @param props the properties of the parameter object
+    */
+  sealed abstract class ImmutableParam(namespace: String, val props: CypherProps)(implicit showProps: Show[CypherProps])
+    extends Param(namespace) with Proxy {
+    final override def isMutable: Boolean = false
+    override def self: Any = (namespace, props)
+    override def toString: String = s"${__clsName}(namespace = $namespace, props = ${showProps show props})"
+  }
+
+  /**
+    * Allows create a single immutable [[CypherProps]] object within the resulting [[CypherParams]]
+    *
+    * Enables the following string interpolation syntax:
+    *
+    * {{{
+    *   val props = Cypher.params("props", Cypher.props("id" -> 1))
+    *   val stmt = cypher"MATCH (n $props) RETURN n"
+    *   stmt.template // "MATCH (n { props }) RETURN n"
+    *   stmt.parameters // Map("id" -> CypherInt(1))
     * }}}
     *
-    * @param namespace the name of the parameters object as sent to the server
+    * @param namespace the name of the parameter object as sent to the server
+    *
     * @return a dynamic parameters builder for embedding into [[CypherStringContext]] interpolated [[CypherStatement]]s
     */
-  def params(namespace: String): Params = new Params(namespace)
+  class DynamicImmutableParam private[Cypher](namespace: String, props: CypherProps)
+    extends ImmutableParam(namespace, props) with Dynamic {
 
-  class Params private[Cypher] (val namespace: String) extends Dynamic {
-    def applyDynamic[T](name: String)(value: T)(implicit writer: CypherWrites[T, CypherValue]): CypherArg = {
-      new CypherParam(namespace, name, writer.writes(value))
+    final override protected def __clsName: String = "SelectDynamicParam"
```

Comment:
No, it's just the class name. I've just been bitten by using the `getClass.getSimpleName` method in Scala, so I figured I would just use an explicit string. I think it would be faster too.